get_devices/web_app.js

```diff
@@ -0,0 +1,154 @@
+ *  limitations under the License.
+ */
+
+// Espruino devices publish a UART service by default.
+const nordicUARTService = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';
+// The test service defined in device_code.js
+const testService = '0b30acec-193e-11eb-adc1-0242ac120002';
+// The test characteristic defined in device_code.js
+const testCharacteristic = '0b30afd0-193e-11eb-adc1-0242ac120002';
+
+let gattServer = undefined;
+let btDeviceId = null;
+const urlParam = 'bt_device_id=';
+const btDeviceIdUrlParam = window.location.search.substring(window.location.search.indexOf(urlParam) + urlParam.length).split('&')[0];
+
+/**
+ * Load the device code to the Espruino IDE.
+ */
+function loadEspruinoDeviceCode() {
+  fetch('device_code.js').then(response => response.text()).then(data => {
+    let url = 'http://www.espruino.com/webide?code=' + encodeURIComponent(data);
+    window.open(url, '_window');
+  });
+}
+
+function onGattDisconnected(evt) {
+  const device = evt.target;
+  logInfo(`Disconnected from GATT on device ${device.name}.`);
+  assertFalse(gattServer.connected, 'Server connected');
+}
+
+async function startPairing() {
+  clearStatus();
+  logInfo('Starting test');
+
+  $('btn_load_code').disabled = true;
+  $('btn_start_pairing').disabled = true;
+  $('btn_refresh_page').disabled = true;
+  $('btn_check_permission').disabled = true;
+
+  try {
+    const options = {
+      filters: [
+        { services: [nordicUARTService] }
+      ],
+      optionalServices: [testService]
+    };
+
+    logInfo(`Requesting Bluetooth device with service ${testService}`);
+    const device = await navigator.bluetooth.requestDevice(options);
+    logInfo(`Connected to Bluetooh device with id: ${device.id} and name: ${device.name}`);
+    btDeviceId = device.id;
+
+    device.addEventListener('gattserverdisconnected', onGattDisconnected);
+    logInfo(`Connecting to GATT server for device \"${device.name}\"...`);
+    gattServer = await device.gatt.connect();
+    assertEquals(gattServer.device, device, 'Server device mismatch');
+    assertTrue(gattServer.connected, 'server.connected should be true');
+
+    logInfo(`Connected to GATT, requesting service: ${testService}...`);
+    const service = await gattServer.getPrimaryService(testService);
+    assertEquals(service.device, device, 'service device mismatch');
+
+    logInfo(`Connected to service uuid: ${service.uuid}, primary:${service.isPrimary}`);
+  } catch (error) {
+    logError(`Unexpected failure: ${error}`);
+  }
+
+  $('btn_load_code').disabled = false;
+  $('btn_start_pairing').disabled = false;
+  $('btn_refresh_page').disabled = false;
+  $('btn_check_permission').disabled = false;
+
+  if (gattServer) {
+    logInfo('Disconnecting from GATT.');
+    gattServer.disconnect();
+  }
+}
+
+async function refreshPage() {
+  logInfo('Refreshing the page');
+  window.location.replace(`${window.location.href.split("?")[0]}?bt_device_id=${btDeviceId}`);
```

Comment:
You can use [`URLSearchParams.toString()`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/toString) to build a new URL.